### ISSUE TITLE
Access to Git server through SSH

### ISSUE THREAD

**Issue by rickpapo** [microsoft/vscode#121839]
Issue Type: <b>Bug</b>

I have a Synology NAS device on my network, and can use Git from the command line against the respository "ssh://user@machine/volume1/git/reponame".  But when I attempt to add a remote server through Visual Studio Code, I get the following output:

> git remote add ssh-//user@machine/volume1/git/reponame ssh://user@machine/volume1/git/reponame
fatal: 'ssh-//user@machine/volume1/git/reponame' is not a valid remote name

Note that the remote server protocol ID was changed from "ssh:" to "ssh-".

Again, I have no problem using the server from a command prompt.  Only Visual Studio Code is having problems.

VS Code version: Code 1.55.2 (3c4e3df9e89829dce27b7b5c24508306b151f30d, 2021-04-13T09:37:02.931Z)
OS version: Linux x64 5.8.0-50-generic snap

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-9700 CPU @ 3.00GHz (2 x 3000)|
|GPU Status|2d_canvas: unavailable_software<br>gpu_compositing: disabled_software<br>multiple_raster_threads: disabled_off<br>oop_rasterization: disabled_off<br>opengl: disabled_off<br>protected_video_decode: disabled_off<br>rasterization: disabled_software<br>skia_renderer: enabled_on<br>video_decode: disabled_software<br>vulkan: disabled_off<br>webgl: unavailable_software<br>webgl2: unavailable_software|
|Load (avg)|0, 0, 0|
|Memory (System)|7.75GB (2.22GB free)|
|Process Argv|--force-user-env --no-sandbox --unity-launch --no-sandbox --crash-reporter-id 79dc7d25-1786-48ef-af1b-7c76cbd7922a|
|Screen Reader|no|
|VM|100%|
|DESKTOP_SESSION|ubuntu|
|XDG_CURRENT_DESKTOP|Unity|
|XDG_SESSION_DESKTOP|ubuntu|
|XDG_SESSION_TYPE|x11|
</details><details><summary>Extensions (20)</summary>

Extension|Author (truncated)|Version
---|---|---
code-gnu-global|aus|0.2.2
doxdocgen|csc|1.2.0
c-cpp-compile-run|dan|1.0.14
msys2|fou|0.7.0
vscode-pull-request-github|Git|0.25.1
better-cpp-syntax|jef|1.15.5
csharp|ms-|1.23.11
python|ms-|2021.4.765268190
jupyter|ms-|2021.5.745244803
remote-containers|ms-|0.166.1
remote-ssh|ms-|0.65.4
remote-ssh-edit|ms-|0.65.4
remote-wsl|ms-|0.54.6
cmake-tools|ms-|1.7.0
cpptools|ms-|1.3.1
cpptools-extension-pack|ms-|1.0.0
vsliveshare|ms-|1.0.4131
vsliveshare-audio|ms-|0.1.91
python|tht|0.2.3
cmake|twx|0.0.17

(1 theme extensions excluded)

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368:30146709
vsreu685:30147344
python383:30185418
pythonvspyt602:30291499
vspor879:30202332
vspor708:30202333
vspor363:30204092
vstry914:30276682
pythonvsdeb440:30248342
pythonvsded773:30248341
pythonvspyt875:30259475
pythonvspyt639:30291489
pythontb:30283811
pythonvspyt551cf:30291415
vspre833cf:30267465
pythonptprofiler:30281270
vshan820:30276952
vscoreces:30290705
pythondataviewer:30285071
vscus158:30286553

```

</details>

**Comment by rickpapo** [comment]
There are two problems with the git command generated by Visual Studio Code:

(1) The ":" after "ssh" is transformed into a dash.
(2) The "name" and "URL" parameters are the same, both of them having been set to the (now invalid) URL.  The command-line "git" program is rejecting the format of the name.  Apparently it wants something similar than the URL.

If I enter "git remote add ssh://user@server/volume1/git/reponame ssh://user@server/volume1/git/reponame" at the command line, it gives this error message:

fatal: 'ssh://user@server/volume1/git/reponame' is not a valid remote name

In other words, it appears the issue to be that the URL is repeated twice, rather than having the Add Remote Repository command prompt separately (or generate) a local name for the remote repository connection.

**Comment by rickpapo** [comment]
If the logic was "let's replace that colon with something harmless", then the change isn't sufficient to make "git" happy.

**Comment by rickpapo** [comment]
The basic problem here appears to be that the alias name for the remote repository (git remote add alias url) is constructed from the URL, and the SSH version of the URL does not get transformed sufficiently so that it is acceptable as an alias name.  This may have to do with the '@' character used to separate the remote user name from the remote server name.

**Comment by rickpapo** [comment]
It would appear that the function "sanitizeRemoteName" in extensions\git\src\commands.ts needs to be altered.

return name && name.replace(/^\.|\/\.|\.\.|~|\^|:|\/$|\.lock$|\.lock\/|\\|\*|\s|^\s*$|\.$|\[|\]$/g, '-');

This needs to be altered so that @ characters also get replaced by hyphens.

**Comment by eamodio** [comment]
@rickpapo Would you mind opening a PR for that?

**Comment by rickpapo** [comment]
I am not familiar with the jargon of GitHub.  What do you mean by “PR”?

**Comment by rickpapo** [comment]
And, in any case, there are other problems in VSCode related to using SSH as the Git connection protocol.  The logic seems to assume you are using HTTPS all the time.
